core/dnsserver/server_quic.go

```diff
@@ -0,0 +1,293 @@
+	DoQCodeProtocolError quic.ApplicationErrorCode = 1
+)
+
+// ServerQUIC represents an instance of a DNS-over-QUIC server.
+type ServerQUIC struct {
+	*Server
+	listenAddr   net.Addr
+	tlsConfig    *tls.Config
+	quicConfig   *quic.Config
+	quicListener *quic.Listener
+
+	// bytesPool is a pool of byte slices used to read DNS packets.
+	bytesPool *sync.Pool
+}
+
+// NewServerQUIC returns a new CoreDNS QUIC server and compiles all plugin in to it.
+func NewServerQUIC(addr string, group []*Config) (*ServerQUIC, error) {
+	s, err := NewServer(addr, group)
+	if err != nil {
+		return nil, err
+	}
+	// The *tls* plugin must make sure that multiple conflicting
+	// TLS configuration returns an error: it can only be specified once.
+	var tlsConfig *tls.Config
+	for _, z := range s.zones {
+		for _, conf := range z {
+			// Should we error if some configs *don't* have TLS?
+			tlsConfig = conf.TLSConfig
+		}
+	}
+
+	if tlsConfig != nil {
+		tlsConfig.NextProtos = []string{"doq"}
+	}
+
+	bytesPool := &sync.Pool{
+		New: func() interface{} {
+			// 2 bytes may be used to store packet length (see TCP/TLS)
+			b := make([]byte, 2+dns.MaxMsgSize)
+			return &b
+		},
+	}
+
+	var quicConfig *quic.Config
+	quicConfig = &quic.Config{
+		MaxIdleTimeout:        s.idleTimeout,
+		MaxIncomingStreams:    math.MaxUint16,
+		MaxIncomingUniStreams: math.MaxUint16,
+		// Enable 0-RTT by default for all connections on the server-side.
+		Allow0RTT: true,
+	}
+
+	return &ServerQUIC{Server: s, tlsConfig: tlsConfig, quicConfig: quicConfig, bytesPool: bytesPool}, nil
+}
+
+// ServePacket implements caddy.UDPServer interface.
+func (s *ServerQUIC) ServePacket(p net.PacketConn) error {
+	s.m.Lock()
+	s.listenAddr = s.quicListener.Addr()
+	s.m.Unlock()
+
+	return s.ServeQUIC()
+}
+
+// ServeQUIC listens for incoming QUIC packets.
+func (s *ServerQUIC) ServeQUIC() error {
+	for {
+		conn, err := s.quicListener.Accept(context.Background())
+		if err != nil {
+			closeQUICConn(conn, DoQCodeNoError)
+			return err
+		}
+
+		go s.serveQUICConnection(conn)
+	}
+}
+
+// serveQUICConnection handles a new QUIC connection. It waits for new streams
+// and passes them to serveQUICStream.
+func (s *ServerQUIC) serveQUICConnection(conn quic.Connection) {
+	for {
+		// In DoQ, one query consumes one stream.
+		// The client MUST select the next available client-initiated bidirectional
+		// stream for each subsequent query on a QUIC connection.
+		stream, err := conn.AcceptStream(context.Background())
+		if err != nil {
+			closeQUICConn(conn, DoQCodeNoError)
+			return
+		}
+
+		go func() {
+			s.serveQUICStream(stream, conn)
+
+			// The server MUST send the response(s) on the same stream and MUST
+			// indicate, after the last response, through the STREAM FIN
+			// mechanism that no further data will be sent on that stream.
+			_ = stream.Close()
+		}()
+
+	}
+}
+
+func (s *ServerQUIC) serveQUICStream(stream quic.Stream, conn quic.Connection) {
+	bufPtr := s.bytesPool.Get().(*[]byte)
```

Comment:
I like the usage of pooled resource, but it hides the fact that we are allocating memory buffer 65535+2 bytes for message regardless of actual size of DNS message, that seems like not efficient usage of memory, can't we rather:
* read 2-byte [prefix](https://github.com/coredns/coredns/pull/6182/files#diff-d4ed07e777c92c5e42daca148e60fbb401f40fa8ba727323ed782e6a4012fcf0R32) to get DNS message length
* allocate buffer according to DNS message length
* read rest of the stream into the buffer?

Reply:
I agree. The problem with this approach is that (at least for sync.Pool) all buffers must be of the same size.
We could use the following library that offers a sized pool of buffers: https://github.com/libp2p/go-buffer-pool. What do you think?

Reply:
imho I would keep it simple for now and ditch the resource pool, it is a premature resource optimization